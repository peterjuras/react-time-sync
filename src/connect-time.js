import React, { Component } from 'react';

import PropTypes from 'prop-types';

const DEFAULT_COMPONENT_CONFIG = {
  timeProp: 'currentTime'
};

function validateComponentConfig(componentConfig) {
  if (
    typeof componentConfig.timeProp !== 'undefined' &&
    (
      typeof componentConfig.timeProp !== 'string' ||
      componentConfig.timeProp === ''
    )
  ) {
    throw new Error('timeProp must be a non-empty string value');
  }
}

export function connectTime(timerConfig, componentConfig = {}) {
  validateComponentConfig(componentConfig);

  const usedTimerConfig = timerConfig || {};
  const usedComponentConfig = {
    ...DEFAULT_COMPONENT_CONFIG,
    ...componentConfig
  };

  return (WrappedComponent) => {
    class TimeComponent extends Component {
      static contextTypes = {
        timeSync: PropTypes.object
      };

      constructor(props, context) {
        super(props);

        if (!context.timeSync) {
          throw new Error('Warning! TimeSync context cannot be found. Did you add <TimeProvider /> at the top of your component hierarchy?');
        }

        this.state = {
          [usedComponentConfig.timeProp]: context.timeSync.getCurrentTime(usedTimerConfig)
        };
      }

      componentDidMount() {
        this.removeTimer = this.context.timeSync.addTimer(this.onTick, usedTimerConfig);
      }

      componentWillUnmount() {
        if (this.removeTimer) {
          this.removeTimer();
        }
      }

      onTick = (currentTime) => {
        this.setState({
          [usedComponentConfig.timeProp]: currentTime
        });
      };

      render() {
        return <WrappedComponent {...this.props} {...this.state} />;
      }
    }

    return TimeComponent;
  };
}
