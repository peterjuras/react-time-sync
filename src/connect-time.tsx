import React from "react";
import { ReactComponentLike } from "prop-types";
import { useTime, TimerConfig } from "./use-time";

interface ComponentConfig {
  timeProp: string;
}

const DEFAULT_COMPONENT_CONFIG: ComponentConfig = {
  timeProp: "currentTime",
};

function validateComponentConfig(
  componentConfig: Partial<ComponentConfig>,
): void {
  if (
    typeof componentConfig.timeProp !== "undefined" &&
    (typeof componentConfig.timeProp !== "string" ||
      componentConfig.timeProp === "")
  ) {
    throw new Error("timeProp must be a non-empty string value");
  }
}

export function connectTime(
  timerConfig?: TimerConfig | null,
  componentConfig: Partial<ComponentConfig> = {},
): (
  WrappedComponent: ReactComponentLike,
) => (props: Record<string, unknown>) => JSX.Element {
  validateComponentConfig(componentConfig);

  const usedTimerConfig = timerConfig || {};
  const usedComponentConfig = {
    ...DEFAULT_COMPONENT_CONFIG,
    ...componentConfig,
  };

  return (
    WrappedComponent: ReactComponentLike,
  ): ((props: Record<string, unknown>) => JSX.Element) => {
    return function TimeComponent(props: Record<string, unknown>): JSX.Element {
      const timeProps = {
        ...props,
        [usedComponentConfig.timeProp]: useTime(usedTimerConfig),
      };
      return <WrappedComponent {...timeProps} />;
    };
  };
}
