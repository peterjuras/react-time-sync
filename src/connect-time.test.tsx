import React from "react";
import { connectTime } from "./connect-time";
import { createMockProvider } from "../test/mock-time-provider";
import FakeTimers from "@sinonjs/fake-timers";
import { act, render, cleanup } from "@testing-library/react";

describe("#connectTime", () => {
  let clock: FakeTimers.Clock;

  beforeEach(() => {
    // @ts-expect-error install types are too strict, see https://github.com/sinonjs/fake-timers/issues/382
    clock = FakeTimers.install({ now: 1 });
  });

  afterEach(() => {
    clock.uninstall();
    cleanup();
  });

  it("should be exported correctly", () => {
    expect(connectTime).toBeInstanceOf(Function);
  });

  it("should return wrapped component without any arguments", () => {
    expect(connectTime()).toBeDefined();
  });

  it("should throw for invalid timeProp", () => {
    expect(() =>
      connectTime(null, { timeProp: 23 } as any)
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      connectTime(null, { timeProp: 23.6 } as any)
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      connectTime(null, { timeProp: [] } as any)
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      connectTime(null, { timeProp: ["a"] } as any)
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      connectTime(null, { timeProp: {} } as any)
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      connectTime(null, { timeProp: { b: 1 } } as any)
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      connectTime(null, { timeProp: "" })
    ).toThrowErrorMatchingSnapshot();
  });

  it("should pass through timerConfig to addTimer", () => {
    const addTimer = jest.fn();
    const MockProvider = createMockProvider({
      addTimer,
    });
    const timerConfig = {
      unit: 1,
    };
    const WrappedComponent = connectTime(timerConfig)(() => <div />);

    render(
      <MockProvider>
        <WrappedComponent />
      </MockProvider>
    );

    expect(addTimer).toHaveBeenCalledTimes(1);
    expect(addTimer.mock.calls[0][1]).toBe(timerConfig);
  });

  it("should call removeTimer on componentWillUnmount", () => {
    const removeTimer = jest.fn();
    const addTimer = jest.fn(() => removeTimer);
    const MockProvider = createMockProvider({
      addTimer,
    });
    const timerConfig = {
      unit: 1,
    };
    const WrappedComponent = connectTime(timerConfig)(() => <div />);

    const { unmount } = render(
      <MockProvider>
        <WrappedComponent />
      </MockProvider>
    );
    expect(removeTimer).toHaveBeenCalledTimes(0);
    unmount();
    expect(removeTimer).toHaveBeenCalledTimes(1);
  });

  it("should update children on a timer tick", () => {
    const WrappedComponent = connectTime()(({ currentTime }) => (
      <div>{currentTime}</div>
    ));

    const { asFragment } = render(<WrappedComponent />);

    expect(asFragment()).toMatchSnapshot();
    act(() => {
      clock.tick(999);
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("should use the specified property name for timeProp", () => {
    const EmptyComponent = jest.fn<
      ReturnType<React.FC>,
      Parameters<React.FC<{ test1: number }>>
    >(() => null);
    const WrappedComponent = connectTime(null, { timeProp: "test1" })(
      EmptyComponent
    );

    render(<WrappedComponent />);

    expect(EmptyComponent).toHaveBeenLastCalledWith({ test1: 0 }, {});
    act(() => {
      clock.tick(999);
    });
    expect(EmptyComponent).toHaveBeenLastCalledWith({ test1: 1 }, {});
  });

  it("should be reusable for multiple components", () => {
    const connect = connectTime();
    const Child1 = connect(({ currentTime }) => (
      <div>Child 1: {currentTime}</div>
    ));
    const Child2 = connect(({ currentTime }) => (
      <div>Child 2: {currentTime}</div>
    ));
    const Child3 = connect(({ currentTime }) => (
      <div>Child 3: {currentTime}</div>
    ));
    const { asFragment } = render(
      <>
        <Child1 />
        <Child2 />
        <Child3 />
      </>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
