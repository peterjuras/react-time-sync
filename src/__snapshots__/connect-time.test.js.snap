// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#connectTime should be reusable for multiple components 1`] = `
<TimeProvider>
  <TimeComponent>
    <Component
      currentTime={0}
    >
      <div>
        Child 1: 
        0
      </div>
    </Component>
  </TimeComponent>
  <TimeComponent>
    <Component
      currentTime={0}
    >
      <div>
        Child 2: 
        0
      </div>
    </Component>
  </TimeComponent>
  <TimeComponent>
    <Component
      currentTime={0}
    >
      <div>
        Child 3: 
        0
      </div>
    </Component>
  </TimeComponent>
</TimeProvider>
`;

exports[`#connectTime should throw for invalid timeProp 1`] = `"timeProp must be a non-empty string value"`;

exports[`#connectTime should throw for invalid timeProp 2`] = `"timeProp must be a non-empty string value"`;

exports[`#connectTime should throw for invalid timeProp 3`] = `"timeProp must be a non-empty string value"`;

exports[`#connectTime should throw for invalid timeProp 4`] = `"timeProp must be a non-empty string value"`;

exports[`#connectTime should throw for invalid timeProp 5`] = `"timeProp must be a non-empty string value"`;

exports[`#connectTime should throw for invalid timeProp 6`] = `"timeProp must be a non-empty string value"`;

exports[`#connectTime should throw for invalid timeProp 7`] = `"timeProp must be a non-empty string value"`;

exports[`#connectTime should throw if context is not found 1`] = `"Warning! TimeSync context cannot be found. Did you add <TimeProvider /> at the top of your component hierarchy?"`;

exports[`#connectTime should update children on a timer tick 1`] = `
<TimeProvider>
  <TimeComponent>
    <Component
      currentTime={0}
    >
      <div>
        0
      </div>
    </Component>
  </TimeComponent>
</TimeProvider>
`;

exports[`#connectTime should update children on a timer tick 2`] = `
<TimeProvider>
  <TimeComponent>
    <Component
      currentTime={1}
    >
      <div>
        1
      </div>
    </Component>
  </TimeComponent>
</TimeProvider>
`;
